src/Chapter3.hs

```diff
@@ -480,6 +532,41 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Building = Church | Library
+
+{-
+params:	Int - population
+		Building - either church or library
+		String - name of castle
+		Bool - if wall is present
+-}
+data City
+	= SmallCity Int Building
+	| CityOfWizards Int Building String Bool
+
+buildCastle :: City -> String -> City
+buildCastle c name = case c of
+	SmallCity p b -> CityOfWizards p b name False
+	CityOfWizards p b _ w -> CityOfWizards p b name w
+
+buildWalls :: City -> City
+buildWalls c = case c of
+	SmallCity _ _ -> error "Build a castle before building walls"
+	CityOfWizards p b n w ->
+		if (p >= 10)
+		then
+			if length n > 0
+			then
+				CityOfWizards p b n True
+			else
+				error "Build a castle before building walls"
+		else
+			error "Build more houses before building walls"
+
+buildHouse :: City -> City
+buildHouse c = case c of
+	SmallCity p b -> SmallCity (p+1) b
+	CityOfWizards p b n w -> CityOfWizards (p+2) b n w
```

Comment:
According to our task description, houses can contain from 1 to 4 people each. But in your solution it looks like each house is either for a single person or for two people :thinking: